stdlib/array.ncl

```diff
@@ -108,12 +108,17 @@
             6
         ```
         "%
-      = fun f acc l =>
-        if %length% l == 0 then
-          acc
+    = fun f acc l =>
+        let length = %length% l in
+        if length == 0 then acc
         else
-          let newAcc = f acc (%head% l) in
-          %seq% newAcc (foldl f newAcc (%tail% l)),
+          let rec loop = fun acc n => 
+            if n == length then acc
+            else 
+              let next_acc = %elem_at% l n |> f acc in
+              %seq% next_acc (loop next_acc (n+1))
+          in
+          loop acc 0,
```

Comment:
I don't really have a preference for either `go` or `loop` but we should probably use the same naming style for both functions.

```suggestion
          let rec go = fun acc n => 
            if n == length then acc
            else 
              let next_acc = %elem_at% l n |> f acc in
              %seq% next_acc (go next_acc (n+1))
          in
          loop acc 0,
```